elastic/kibana
# Instrument vis_type_vislib, lens and vis_type_timeseries with execution context service

## Summary

Part of https://github.com/elastic/kibana/issues/101587
It integrates a few visualizations with a mechanism allowing users to understand what Kibana entity initiated a search request.
It can be used to identify a source of slowness via Elasticsearch [slow log query](https://www.elastic.co/guide/en/elasticsearch/reference/master/index-modules-slowlog.html):  the slow log record contains `x-opaque-id` header, which value is propagated from the Kibana server or browser App.

The current PR integrates the `execution_context` service with :
- vis_type_vislib https://github.com/elastic/kibana/blob/cc4157110cf58d99e693e4d0800abf2e82ddd31f/src/plugins/vis_type_vislib/public/types.ts#L39-L50
- lens
- vis_type_timeseries

I'd appreciate any suggestion from the team-owners on improving the logic or general guidance on the `execution_context` service integration.

### Implementation
The browser environment doesn't provide an API to keep track of the "thread" context during async operations.
In nodejs env, we rely on [AsyncLocalStorage](https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage) and [Async hooks](https://nodejs.org/api/async_hooks.html) for this purpose. 
Thus, for the browser env, we have to implement an alternative approach. A Kibana plugin creates an `async context object` and passes it manually through the whole invocation chain to a place when the Kibana browser app communicates with the Kibana server via an HTTP protocol. Then the propagated `async context object` is serialized and passed through the network to the Kibana server. In turn, the Kibana server stores it as a part of the "thread" context, emits `async context object` to the logs, and attaches id and type of `async context object` as part of `x-opaque-id` header to every call to the Elasticsearch server. Therefore, a Kibana user can identify what a feature initiated a request via Kibana logs (under `elasticsearch.query` logger) or Elasticsearch logs (`x-opaque-id` is attached to `search slow logs`).

### How to test
Unfortunately, right at this moment, only manually. I'm going to add integration tests in a follow-up PR.
Steps to test the PR manually:
- setup logging to observe `request.id: opaqueId` in `elasticsearch response` and `execution_context` whenever it's set
```yml
logging:
  appenders:
    myconsole:
      type: console
      layout:
        type: pattern
        pattern: '%date|%meta'
  loggers:
    - name: elasticsearch.query
      level: all
      appenders: [myconsole]
    - name: elasticsearch.query.taskManager
      level: off
    - name: execution_context
      level: debug
      appenders: [console]
```
- run Kibana
- navigate to Kibana Browser App
- Install sample data. I tested on `Sample eCommerce orders`.
- Navigate to dashboards.
- You should be able to see the output like this:
```
[2021-07-12T13:14:28.944+03:00][DEBUG][execution_context] stored the execution context: {"requestId":"d568e29f-c3de-4ef6-a311-225b1358f8b7","type":"visuzalization","name":"gauge","description":"[eCommerce] Average Sales Price","id":"4b3ec120-b892-11e8-a6d9-e546fe2bba5f","url":"/bon/app/visualize#/edit/4b3ec120-b892-11e8-a6d9-e546fe2bba5f"}

2021-07-12T13:14:28.965+03:00|{"http":{"request":{"id":"d568e29f-c3de-4ef6-a311-225b1358f8b7;kibana:visuzalization:gauge:4b3ec120-b892-11e8-a6d9-e546fe2bba5f"}}}
```
